### PR TITLE
eliminate useless dynamic classes, store reference to folded componentStyle

### LINKED ISSUE #3206: Complex selectors with ampersands break after the component is extended.
## Reproduction
https://codepen.io/eduardakhmetshin/pen/ZEQmyNd

## Steps to reproduce
1. Create a styled component with a complex selector with ampersands. For example:
`& + & {}`. Every ampersand in the rule corresponds to a different class that is assigned to the component.
2. Extend this component.

## Expected Behavior
All classes remain for extending components and complex rules with ampersands don't break. In this example, all divs after the first should have red borders.

## Actual Behavior
Some classes are dropped for extending components which breaks rules. In this example, the last div remains black.

### PR DESCRIPTION
**Pull request by probablyup**
Fixes #3206 

1. this effectively deduplicates a lot of redundant CSS that is generated
   starting with v4

2. & self-references are now correctly resolved to the folded layer they
   correspond to

### CODE REVIEW

**Comment by probablyup** [comment]
Published as styled-components@5.2.0-test.0 for testing

**Comment by probablyup** [comment]
This doesn't quite fix #3206 yet, just realized the selectors for composed self-references need to be based on the OG static class name...

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "recommendations": [
+    "christian-kohler.path-intellisense",
```
Comment (kitten) [inline .vscode/extensions.json:3]: Accidentaly change?

Comment (probablyup) [reply .vscode/extensions.json:3]: No, just decided to add it in while I Was doing other stuff 😅

File: packages/styled-components/src/models/StyledComponent.js [hunk 2]
```diff
@@ -225,12 +225,9 @@ export default function createStyledComponent(
...
   const componentStyle = new ComponentStyle(
-    isTargetStyledComp
-      ? // fold the underlying StyledComponent rules up (implicit extend)
-        // $FlowFixMe
-        target.componentStyle.rules.concat(rules)
-      : rules,
-    styledComponentId
+    rules,
+    styledComponentId,
+    isTargetStyledComp ? ((target: Object).componentStyle: ComponentStyle) : undefined
```
Comment (kitten) [inline packages/styled-components/src/models/StyledComponent.js:230]: Super nice!

**Review by kitten — APPROVED** [review]

**Comment by Zenser** [comment]
@probablyup hi, i see this #2071 use for style isolate.
```css
// Comp -> .sc-a
// color: red -> .a
// color: blue -> .b

.sc-a {}
.a { color: red }
.a + .sc-a { margin-left: 10px }
.b { color: blue }
.b + .sc-a { margin-left: 10px }
```
It is useful for me, and now how do style isolate?

Has many styled-components instants in one container at my case. Thanks for your callback.

**Pull request merged**